framework/base/Application.php

```diff
@@ -456,6 +468,88 @@ public function setVendorPath($path)
     }
 
     /**
+     * Registers a class definitions in [[yii\di\Container|Dependency Injection Container]]
+     * by calling [[yii\di\Container::set()]].
+     * This method is provided so that you can configure Dependency Injection Container
+     * when configuring an application.
+     *
+     * @param array $definitions array of definitions. There are two allowed formats of array:
+     * The first format:
+     *  - key: class name, interface name or alias name. The key will be passed to the
+     * [[yii\di\Container::set()]] method as a first argument `$class`.
+     *  - value: the definition associated with `$class`. Possible values are described in
+     * [[yii\di\Container::set()]] documentation for the `$definition` parameter. Will be passed to
+     * the [[yii\di\Container::set()]] method as the second argument `$definition`.
+     *
+     * Example:
+     * ```php
+     * [
+     *     'yii\web\Request' => 'app\components\Request',
+     *     'yii\web\Response' => [
+     *         'class' => 'app\components\Response',
+     *         'format' => 'json'
+     *     ],
+     *     'foo\Bar' => function () {
+     *         $qux = new Qux;
+     *         $foo = new Foo($qux);
+     *         return new Bar($foo);
+     *     }
+     * ]
+     * ```
+     *
+     * The second format:
+     *  - key: class name, interface name or alias name. The key will be passed to the
+     * [[yii\di\Container::set()]] method as a first argument `$class`.
+     * - value: array of two elements. The first element will be passed the [[yii\di\Container::set()]]
+     * method as the second argument `$definition`, the second one - as `$params`.
+     *
+     * Example:
+     * ```php
+     * [
+     *     'foo\Bar' => [
+     *          ['class' => 'app\Bar'],
+     *          [Instance::of('baz')]
+     *      ]
+     * ]
+     * ```
+     *
+     * @see \yii\di\Container::set() to know more about possible values of definitions
+     */
+    public function setContainerDefinitions(array $definitions)
```

Comment:
https://github.com/yiisoft/yii2/pull/13029#issuecomment-261243153

See @samdark's comment. I agree with it. Do you?